File: src/test/java/uk/gov/justice/hmpps/prison/util/QueryBuilderTest.java
package uk.gov.justice.hmpps.prison.util;


import com.google.common.collect.ImmutableMap;
import com.tngtech.java.junit.dataprovider.DataProvider;
import com.tngtech.java.junit.dataprovider.DataProviderRunner;
import com.tngtech.java.junit.dataprovider.UseDataProvider;
import org.junit.Test;
import org.junit.runner.RunWith;
import uk.gov.justice.hmpps.prison.repository.mapping.FieldMapper;

import java.util.Map;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static uk.gov.justice.hmpps.prison.util.DatabaseDialect.HSQLDB;
import static uk.gov.justice.hmpps.prison.util.DatabaseDialect.ORACLE_11;
import static uk.gov.justice.hmpps.prison.util.DatabaseDialect.ORACLE_12;
import static uk.gov.justice.hmpps.prison.util.DatabaseDialect.POSTGRES;

@RunWith(DataProviderRunner.class)
public class QueryBuilderTest {

    private final Map<String, FieldMapper> agencyMapping = new ImmutableMap.Builder<String, FieldMapper>()
            .put("AGENCY_CODE", new FieldMapper("agencyId"))
            .put("DESCRIPTION", new FieldMapper("description")
            ).build();

    private static final String FROM_AGENCY_LOCATIONS = "FROM AGENCY_LOCATIONS A";
    private static final String INITIAL_SQL = "SELECT A.AGY_LOC_ID AGENCY_CODE, A.DESCRIPTION " + FROM_AGENCY_LOCATIONS;

    private static final QueryBuilderFactory oracle11Builder = new QueryBuilderFactory(ORACLE_11);
    private static final QueryBuilderFactory oracle12Builder = new QueryBuilderFactory(ORACLE_12);
    private static final QueryBuilderFactory postgresBuilder = new QueryBuilderFactory(POSTGRES);
    private static final QueryBuilderFactory hsqldbBuilder = new QueryBuilderFactory(HSQLDB);

    @DataProvider
    public static Object[][] data() {
        return new Object[][]{
                {"eq", "=", "'ITAG'", "'ITAG'"},
                {"neq", "!=", "'ITAG'", "'ITAG'"},
                {"is", "IS", "NULL", "NULL"},
                {"gt", ">", "'ITAG'", "'ITAG'"},
                {"gteq", ">=", "'ITAG'", "'ITAG'"},
                {"lt", "<", "'ITAG'", "'ITAG'"},
                {"lteq", "<=", "'ITAG'", "'ITAG'"},
                {"like", "LIKE", "'ITA%'", "'ITA%'"},
                {"in", "IN", "'X'|'B'", "('X','B')"}
        };
    }


    @Test
    public void buildSqlPreOracle12WithoutPaginationShouldReturnTheSameSQL() {
        final var sql = oracle11Builder.getQueryBuilder(INITIAL_SQL, agencyMapping).removeSpecialChars().build();
        assertThat(sql, equalTo(INITIAL_SQL));
    }

    @Test
    public void buildSqlOracle12WithoutPaginationShouldReturnTheSameSql() {
        final var sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping).removeSpecialChars().build();
        assertThat(sql, equalTo(INITIAL_SQL));
    }

    @Test
    public void buildSqlOracle12AndRowCountShouldReturnOracleNativePaginatedSql() {
        final var sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addRowCount()
                .build();
        final var expectedSql = "SELECT COUNT(*) OVER() RECORD_COUNT, QRY_ALIAS.* FROM ( " + INITIAL_SQL + " ) QRY_ALIAS";
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlPostgresAndRowCountShouldReturnOracleNativePaginatedSql() {
        final var sql = postgresBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addRowCount()
                .build();
        final var expectedSql = "SELECT COUNT(*) OVER() RECORD_COUNT, QRY_ALIAS.* FROM ( " + INITIAL_SQL + " ) QRY_ALIAS";
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlHsqlDBAndRowCountShouldReturnOracleNativePaginatedSql() {
        final var sql = hsqldbBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addRowCount()
                .build();
        final var expectedSql = "WITH TOTAL_COUNT AS ( SELECT COUNT(*) AS RECORD_COUNT FROM (" + INITIAL_SQL + ") QRY_ALIAS ) SELECT * FROM TOTAL_COUNT, (SELECT QRY_ALIAS.* FROM (" + INITIAL_SQL + ") QRY_ALIAS";
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlOracle12AndPaginationShouldReturnOracleNativePaginatedSql() {
        final var sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addPagination()
                .build();
        final var expectedSql = INITIAL_SQL + " OFFSET :offset ROWS FETCH NEXT :limit ROWS ONLY";
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlPostgresAndPaginationShouldReturnOracleNativePaginatedSql() {
        final var sql = postgresBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addPagination()
                .build();
        final var expectedSql = INITIAL_SQL + " OFFSET (:offset) ROWS FETCH NEXT (:limit) ROWS ONLY";
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlHsqlDBAndPaginationShouldReturnOracleNativePaginatedSql() {
        final var sql = hsqldbBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addPagination()
                .build();
        final var expectedSql = INITIAL_SQL + "  OFFSET :offset ROWS FETCH NEXT :limit ROWS ONLY";
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlPreOracle12AndPaginationShouldReturnSubQueryPaginatedSql() {
        final var sql = oracle11Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addPagination()
                .build();
        final var expectedSql = String.format("SELECT * FROM (SELECT QRY_PAG.*, ROWNUM rnum FROM ( %s  ) QRY_PAG WHERE ROWNUM <= :offset+:limit) WHERE rnum >= :offset+1", INITIAL_SQL);
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlUsingAliasColumnWithConditionShouldReturnQueryWithSubQuery() {
        final var sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addQuery("agencyId:eq:'ITAG'")
                .build();
        final var expectedSql = String.format("SELECT QRY_ALIAS.* FROM ( %s ) QRY_ALIAS WHERE AGENCY_CODE = 'ITAG'", INITIAL_SQL);
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    public void buildSqlUsingAliasColumnWithConditionAndRowCountingShouldReturnQueryWithSubQuery() {
        final var sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addQuery("agencyId:eq:'ITAG'")
                .addRowCount()
                .build();
        final var expectedSql = String.format("SELECT COUNT(*) OVER() RECORD_COUNT, QRY_ALIAS.* FROM ( %s ) QRY_ALIAS WHERE AGENCY_CODE = 'ITAG'", INITIAL_SQL);
        assertThat(sql, equalTo(expectedSql));
    }

    @Test
    @UseDataProvider("data")
    public void buildSqlUsingQueryOperatorsShouldReturnExpectedParam(final String jsonOperator, final String sqlOperator, final String paramValue, final String paramExpected) {
        final var sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
                .removeSpecialChars()
                .addQuery("agencyId:" + jsonOperator + ":" + paramValue)
                .build();
        final var expectedSql = String.format("SELECT QRY_ALIAS.* FROM ( %s ) QRY_ALIAS WHERE AGENCY_CODE %s %s", INITIAL_SQL, sqlOperator, paramExpected);
        assertThat(sql, equalTo(expectedSql));
    }
}

File: src/test/java/uk/gov/justice/hmpps/prison/util/QueryBuilderTest.kt
package uk.gov.justice.hmpps.prison.util

import io.opentelemetry.sdk.testing.assertj.OpenTelemetryAssertions.assertThat
import org.junit.Test
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.Arguments.arguments
import org.junit.jupiter.params.provider.MethodSource
import uk.gov.justice.hmpps.prison.repository.mapping.FieldMapper
import java.util.stream.Stream

class QueryBuilderTest {
  private val agencyMapping = mapOf(
    "AGENCY_CODE" to FieldMapper("agencyId"),
    "DESCRIPTION" to FieldMapper("description"),
  )

  @Test
  fun buildSqlPreOracle12WithoutPaginationShouldReturnTheSameSQL() {
    val sql = oracle11Builder.getQueryBuilder(INITIAL_SQL, agencyMapping).removeSpecialChars().build()
    assertThat(sql).isEqualTo(INITIAL_SQL)
  }

  @Test
  fun buildSqlOracle12WithoutPaginationShouldReturnTheSameSql() {
    val sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping).removeSpecialChars().build()
    assertThat(sql).isEqualTo(INITIAL_SQL)
  }

  @Test
  fun buildSqlOracle12AndRowCountShouldReturnOracleNativePaginatedSql() {
    val sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addRowCount()
      .build()
    val expectedSql = "SELECT COUNT(*) OVER() RECORD_COUNT, QRY_ALIAS.* FROM ( $INITIAL_SQL ) QRY_ALIAS"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlPostgresAndRowCountShouldReturnOracleNativePaginatedSql() {
    val sql = postgresBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addRowCount()
      .build()
    val expectedSql = "SELECT COUNT(*) OVER() RECORD_COUNT, QRY_ALIAS.* FROM ( $INITIAL_SQL ) QRY_ALIAS"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlHsqlDBAndRowCountShouldReturnOracleNativePaginatedSql() {
    val sql = hsqldbBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addRowCount()
      .build()
    val expectedSql =
      "WITH TOTAL_COUNT AS ( SELECT COUNT(*) AS RECORD_COUNT FROM ($INITIAL_SQL) QRY_ALIAS ) SELECT * FROM TOTAL_COUNT, (SELECT QRY_ALIAS.* FROM ($INITIAL_SQL) QRY_ALIAS"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlOracle12AndPaginationShouldReturnOracleNativePaginatedSql() {
    val sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addPagination()
      .build()
    val expectedSql = "$INITIAL_SQL OFFSET :offset ROWS FETCH NEXT :limit ROWS ONLY"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlPostgresAndPaginationShouldReturnOracleNativePaginatedSql() {
    val sql = postgresBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addPagination()
      .build()
    val expectedSql = "$INITIAL_SQL OFFSET (:offset) ROWS FETCH NEXT (:limit) ROWS ONLY"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlHsqlDBAndPaginationShouldReturnOracleNativePaginatedSql() {
    val sql = hsqldbBuilder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addPagination()
      .build()
    val expectedSql = "$INITIAL_SQL  OFFSET :offset ROWS FETCH NEXT :limit ROWS ONLY"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlPreOracle12AndPaginationShouldReturnSubQueryPaginatedSql() {
    val sql = oracle11Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addPagination()
      .build()
    val expectedSql =
      "SELECT * FROM (SELECT QRY_PAG.*, ROWNUM rnum FROM ( $INITIAL_SQL  ) QRY_PAG WHERE ROWNUM <= :offset+:limit) WHERE rnum >= :offset+1"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlUsingAliasColumnWithConditionShouldReturnQueryWithSubQuery() {
    val sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addQuery("agencyId:eq:'ITAG'")
      .build()
    val expectedSql = "SELECT QRY_ALIAS.* FROM ( $INITIAL_SQL ) QRY_ALIAS WHERE AGENCY_CODE = 'ITAG'"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Test
  fun buildSqlUsingAliasColumnWithConditionAndRowCountingShouldReturnQueryWithSubQuery() {
    val sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
      .removeSpecialChars()
      .addQuery("agencyId:eq:'ITAG'")
      .addRowCount()
      .build()
    val expectedSql = "SELECT COUNT(*) OVER() RECORD_COUNT, QRY_ALIAS.* FROM ( $INITIAL_SQL ) QRY_ALIAS WHERE AGENCY_CODE = 'ITAG'"
    assertThat(sql).isEqualTo(expectedSql)
  }

  @Nested
  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
  inner class BuildWithData {
    @ParameterizedTest
    @MethodSource("data")
    fun buildSqlUsingQueryOperatorsShouldReturnExpectedParam(
      jsonOperator: String,
      sqlOperator: String?,
      paramValue: String,
      paramExpected: String?,
    ) {
      val sql = oracle12Builder.getQueryBuilder(INITIAL_SQL, agencyMapping)
        .removeSpecialChars()
        .addQuery("agencyId:$jsonOperator:$paramValue")
        .build()
      val expectedSql = "SELECT QRY_ALIAS.* FROM ( $INITIAL_SQL ) QRY_ALIAS WHERE AGENCY_CODE $sqlOperator $paramExpected"
      assertThat(sql).isEqualTo(expectedSql)
    }

    fun data(): Stream<Arguments> = Stream.of(
      arguments("eq", "=", "'ITAG'", "'ITAG'"),
      arguments("neq", "!=", "'ITAG'", "'ITAG'"),
      arguments("is", "IS", "NULL", "NULL"),
      arguments("gt", ">", "'ITAG'", "'ITAG'"),
      arguments("gteq", ">=", "'ITAG'", "'ITAG'"),
      arguments("lt", "<", "'ITAG'", "'ITAG'"),
      arguments("lteq", "<=", "'ITAG'", "'ITAG'"),
      arguments("like", "LIKE", "'ITA%'", "'ITA%'"),
      arguments("in", "IN", "'X'|'B'", "('X','B')"),
    )
  }

  companion object {
    private const val FROM_AGENCY_LOCATIONS = "FROM AGENCY_LOCATIONS A"
    private const val INITIAL_SQL = "SELECT A.AGY_LOC_ID AGENCY_CODE, A.DESCRIPTION $FROM_AGENCY_LOCATIONS"
    private val oracle11Builder = QueryBuilderFactory(DatabaseDialect.ORACLE_11)
    private val oracle12Builder = QueryBuilderFactory(DatabaseDialect.ORACLE_12)
    private val postgresBuilder = QueryBuilderFactory(DatabaseDialect.POSTGRES)
    private val hsqldbBuilder = QueryBuilderFactory(DatabaseDialect.HSQLDB)
  }
}
